sotodlib/io/load_book.py

```diff
@@ -0,0 +1,223 @@
+            # data.  It is updated after each frame is processed.
+            # Initially, it could be larger than the number of samples
+            # in the frame we're looking at.  Eventually, it could be
+            # zero.
+            start = max(0, samples[0] - i0)
+
+            filename = os.path.join(prefix, f)
+            for frame in spt3g_core.G3File(filename):
+                if frame.type is not spt3g_core.G3FrameType.Scan:
+                    continue
+
+                if stream_id is None:
+                    stream_id = frame['stream_id']
+                    for p in pairs_req:
+                        if p[0] == detset:
+                            det_info[p[1]].append(stream_id)
+
+                # Extract "primary" fields, organize by detset.
+                if primary_fields is None:
+                    primary_fields = list(frame['primary'].names)
+                else:
+                    assert(primary_fields == list(frame['primary'].names))
+
+                if primary_dest is None:
+                    primary_dest = core.AxisManager(aman.samps)
+                    for f in primary_fields:
+                        primary_dest.wrap_new(f, ('samps', ), dtype='int64')
+                    aman['primary'].wrap(detset, primary_dest)
+
+                primary_block = frame['primary'].data
+                for i, f in enumerate(primary_fields):
+                    delta = _extract_1d(primary_block[i], start,
+                                        primary_dest[f], dest_offset)
+
+                # Extract "tes_bias", organize by detset.
+                if tes_fields is None:
+                    tes_fields = list(frame['tes_bias'].names)
+                else:
+                    assert(tes_fields == list(frame['tes_bias'].names))
+
+                if tes_dest is None:
+                    tes_dest = core.AxisManager(aman.samps)
+                    for f in tes_fields:
+                        tes_dest.wrap_new(f, ('samps', ), dtype='int32')
+                    aman['tes_bias'].wrap(detset, tes_dest)
+
+                tes_block = frame['tes_bias'].data
+                for i, f in enumerate(tes_fields):
+                    delta = _extract_1d(tes_block[i], start,
+                                        tes_dest[f], dest_offset)
+
+                # Extract the main signal
+                if not no_signal:
+                    delta = _extract_2d(frame['signal'], start,
+                                        aman['signal'], dest_offset, aman.dets.vals)
+                
+                # How many samples were added to dest?
+                if delta > start:
+                    dest_offset += (delta - start)
+                    start = 0
+                else:
+                    start -= delta
+
+                if dest_offset >= samples[1] - samples[0]:
+                    break
+
+    det_info = core.metadata.ResultSet(
```

Comment:
I'd like to build the `smurf` entry into the det_info here as well. We don't need to follow to functional path in `load_smurf` but it's something used in analysis fairly often. All these get built off the status frame that's in the first file on an g3 session and so should also be at the beginning of every book frame. 

https://github.com/simonsobs/sotodlib/blob/master/sotodlib/io/load_smurf.py#L2381 

The `status` axismanager is also something we'll want to track through, it has things like what filters are applied to the data as we're taking it, etc. 

At level 2 we've been using the `SmurfStatus.from_file( filename)` function to easily make a status object that can use used for all these. I've already seen this work for books.

Reply:
Cool, will work this in.

Reply:
noting that the per-detset organization will be a thing to track here as well.